torch/nn/modules/_functions.py

```diff
@@ -69,8 +78,10 @@ def forward(self, input, weight, bias, running_mean, running_var, eps, momentum,
         self.process_group = process_group
 
         # apply element-wise normalization
-        out = torch.batch_norm_elemt(input, weight, bias, mean, invstd, eps)
-        return out
+        if input.numel() > 0:
+            return torch.batch_norm_elemt(input, weight, bias, mean, invstd, eps)
+        else:
+            return torch.empty_like(input)
```

Comment:
I've tested something similar before but placed this at the beginning of `forward`. Unfortunately placing it at the end has the same issue, the training hangs. I've tried this combined patch (with all the comments included) on real data & cluster. After a few iterations, when we eventually sent empty data to one of the devices the training freezes. I've sshed to all the nodes of the cluster and I see that all the GPUs that received data report 100% utilization while one of them reports 0%. 

Am I right to suggest that this `if` statement messes with the graph and causes issues on the backwards step? Perhaps what we want is for `batch_norm_elemt` to actually handle empty input?

Reply:
>  all the GPUs that received data report 100% utilization while one of them reports 0%.

This means all expect one is waiting for collective communication, but someone one of them didn't join the collective comm. 

> Am I right to suggest that this if statement messes with the graph and causes issues on the backwards step?

Yep, basically the above error prevents that process from joining the collectives. Let me try a fix tonight.